### PR TITLE
add trigger pipe after authentication

### PR DESCRIPTION
**Pull request by ballinette**
This will enable multi-factor authentication with a plugin.

### CODE REVIEW

File: lib/api/controllers/authController.js [hunk 1]
```diff
@@ -79,8 +79,9 @@ class AuthController {
     const strategy = request.input.args.strategy;
 
     return this.kuzzle.passport.authenticate({body: request.input.body, query: request.input.args, original: request}, strategy)
+      .then(userObject => this.kuzzle.pluginsManager.trigger('passport:afterAuthenticate', {strategy, user:userObject}))
```
Comment (ballinette) [inline lib/api/controllers/authController.js]: we create a JSON object with the strategy the user object, to enable the pipe listener to handle different stuff according to the actual strategy if needed.

File: lib/api/controllers/authController.js [hunk 2]
```diff
@@ -79,21 +79,24 @@ class AuthController {
     const strategy = request.input.args.strategy;
 
     return this.kuzzle.passport.authenticate({body: request.input.body, query: request.input.args, original: request}, strategy)
+      .then(userObject => {
```
Comment (ballinette) [inline lib/api/controllers/authController.js]: we emit a JSON object with the strategy the user object, to enable the pipe listener to handle different stuff according to the actual strategy if needed.

Comment (kuzzle) [inline lib/api/controllers/authController.js]: MINOR Remove curly braces and "return" from this arrow function body. [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3524)

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: lib/api/controllers/authController.js [hunk 3]
```diff
@@ -79,21 +79,22 @@ class AuthController {
     const strategy = request.input.args.strategy;
 
     return this.kuzzle.passport.authenticate({body: request.input.body, query: request.input.args, original: request}, strategy)
+      .then(userObject => this.kuzzle.pluginsManager.trigger('auth:strategyAuthenticated', {strategy, content:userObject}))
```
Comment (ballinette) [inline lib/api/controllers/authController.js]: we emit a JSON object with the strategy the user object, to enable the pipe listener to handle different stuff according to the actual strategy if needed.

File: lib/api/controllers/authController.js [hunk 4]
```diff
@@ -79,21 +79,30 @@ class AuthController {
     const strategy = request.input.args.strategy;
 
     return this.kuzzle.passport.authenticate({body: request.input.body, query: request.input.args, original: request}, strategy)
+      .then(content => {
+        // do not trigger the "auth:strategyAutenticated" pipe if the result if not a User object,
```
Comment (AnthonySendra) [inline lib/api/controllers/authController.js]: nitpick: "the result is not a User ..."

File: lib/api/controllers/authController.js [hunk 5]
```diff
@@ -79,21 +79,30 @@ class AuthController {
     const strategy = request.input.args.strategy;
 
     return this.kuzzle.passport.authenticate({body: request.input.body, query: request.input.args, original: request}, strategy)
+      .then(content => {
+        // do not trigger the "auth:strategyAutenticated" pipe if the result is not a User object,
+        // ie. if we are a intermediate step of a multi-step authentication strategy (example: first redirection call for oAuth strategies)
+        if (! (content instanceof User)) {
```
Comment (AnthonySendra) [inline lib/api/controllers/authController.js]: test

**Pull request merged**